### ISSUE TITLE
About the clangarm64 directory in /

### ISSUE THREAD

**Issue by angelog0** [msys2/MSYS2-packages#2440]
I noticed a new `clangarm64` directory in root folder, '/'. What is? Which is the reason it is there?

**Comment by jeremyd2019** [comment]
it is the beginnings of an __experimental__ native arm64 build of MINGW-packages, built using clang (GCC does not yet support Windows arm64)

**Comment by angelog0** [comment]
@jeremyd2019 wrote:

> it is the beginnings of an experimental native arm64 build of MINGW-packages, built using clang (GCC does not yet support Windows arm64)

Would those MINGW-packages work on a system base on a AMD Ryzen 5 2400G processor?

**Comment by 1480c1** [comment]
They are primarily for the Windows On Arm systems, so Surface Pro X etc

**Comment by angelog0** [comment]
@1480c1 wrote:

> They are primarily for the Windows On Arm systems, so Surface Pro X etc

...but will they work on an a system base on a AMD Ryzen 5 2400G processor? Y/N?

**Comment by 1480c1** [comment]
No, AMD Ryzen 5 2400G is an x86_64 processor, not an arm based processor

**Comment by jeremyd2019** [comment]
At this point what i'm playing with is native, so no to both.  There are cross-compilers existing, and it would be possible to package one (in fact, the packaged mingw-w64-clang could probably target arm64, if provided with libraries for the target, including compiler-rt)

**Comment by angelog0** [comment]
Then, why do you install things on a system on which they will never work? 

If `clangarm64` is for packages which work only on an Arm based systems, you should not install/create it on systems which never will use it.

**Comment by jeremyd2019** [comment]
I have a machine that runs 32-bit Windows.  While i686 msys was still supported, it installed the `mingw64` directory even though the packages that would install there would never work on a 32-bit OS.

**Comment by Biswa96** [comment]
> Then, why do you install things on a system on which they will never work?

The installer only make those empty folders. There will be no executable unless a package is installed. This is also true for other environment also (e.g. /mingw32, /clang32 etc.)

If you want to run ARM executables then you have to boot a Windows 10 ARM64 OS in QEMU.

**Comment by lazka** [comment]
These directories used to contain some autotools config files, but I removed those recently. In theory we cold remove them, so that only if a package is installed they start to exist.

Not sure if something else depends on them existing.

thoughts?

**Comment by msink** [comment]
Ideally there should be some way to build `clangarm64` packages on x86_64 host.
Because Windows/ARM64 host is very exotic thing, but consumer devices like Hololens2 do exist.
Is it possible?

**Comment by Biswa96** [comment]
It has been possible since many years, see https://github.com/mstorsjo/llvm-mingw. Recently, msys2 have started to use llvm to build packages. So, it will take time, probably many.

**Comment by msink** [comment]
Well, of cource Clang/LLVM supports cross-compilation froom the very beginning.
I mean - can/will MSYS2 somehow use together compiler/toolchain in `/msys64/clang64` and target sysroot in `/msys64/clangarm64` to generate Windows-ARM64 binaries using Windows-x64 host?

**Comment by Biswa96** [comment]
Yes, it's possible. But generally packages are built with native compiler. Here cross compiler are used only to compile the native compiler. Optionally , one can extract the llvm-mingw released packages in msys2 and add it in `$PATH` and change the target triplet in build command.

**Comment by jeremyd2019** [comment]
> Ideally there should be some way to build `clangarm64` packages on x86_64 host.

Ideally, yes, that would make life easier.  But,

> But generally packages are built with native compiler. Here cross compiler are used only to compile the native compiler. 

The packaging scripts here assume a 'native' build, not a cross-compile.  It would be a pretty major undertaking to make them successfully cross-compile.  That's why I started out with native arm64.
